src/binary_annotator.cpp

```diff
@@ -0,0 +1,647 @@
+
+  // Build the root table, and all else will be referenced from it.
+  BuildTable(root_table_offset, BinarySectionType::RootTable,
+             schema_->root_table());
+
+  // Now that all the sections are built, scan the regions between them and
+  // insert padding bytes that are implied.
+  InsertFinalPadding();
+
+  return sections_;
+}
+
+uint64_t BinaryAnnotator::BuildHeader(const uint64_t offset) {
+  std::vector<BinaryRegion> regions;
+
+  // TODO(dbaileychess): sized prefixed value
+  const uint32_t root_table_offset = GetScalar<uint32_t>(offset);
+  regions.emplace_back(MakeBinaryRegion(
+      offset, sizeof(uint32_t), BinaryRegionType::UOffset, 0, root_table_offset,
+      std::string("offset to root table `") +
+          schema_->root_table()->name()->str() + "`"));
+
+  const uint32_t file_identifier_value =
+      GetScalar<uint32_t>(offset + sizeof(uint32_t));
+
+  if (file_identifier_value) {
+    // Check if the file identifier region has non-zero data, and assume its the
+    // file identifier. Otherwise, it will get filled in with padding later.
+    regions.emplace_back(MakeBinaryRegion(
+        offset + sizeof(uint32_t), 4 * sizeof(uint8_t), BinaryRegionType::Char,
+        4, 0, std::string("File Identifier")));
+  }
+
+  sections_.insert(std::make_pair(
+      offset, MakeBinarySection("", BinarySectionType::Header, regions)));
+  return root_table_offset;
+}
+
+void BinaryAnnotator::BuildVTable(uint64_t offset,
+                                  const reflection::Object *table) {
+  const uint64_t vtable_offset = offset;
+
+  // First see if we have used this vtable before, if so skip building it again.
+  auto it = vtables_.find(vtable_offset);
+  if (it != vtables_.end()) { return; }
+
+  std::vector<BinaryRegion> regions;
+
+  // Vtables start with the size of the vtable
+  const uint16_t vtable_size = GetScalar<uint16_t>(offset);
+  regions.emplace_back(MakeBinaryRegion(offset, sizeof(uint16_t),
+                                        BinaryRegionType::Uint16, 0, 0,
+                                        std::string("size of this vtable")));
+  offset += sizeof(uint16_t);
+
+  // Then they have the size of the table they reference.
+  regions.emplace_back(
+      MakeBinaryRegion(offset, sizeof(uint16_t), BinaryRegionType::Uint16, 0, 0,
+                       std::string("size of referring table")));
+  offset += sizeof(uint16_t);
+
+  const uint64_t offset_start = offset;
+
+  // A mapping between field (and its id) to the relative offset (uin16_t) from
+  // the start of the table.
+  std::vector<std::tuple<const reflection::Field *, uint16_t>> fields;
+
+  // Loop over all the fields.
+  ForAllFields(table, /*reverse=*/false, [&](const reflection::Field *field) {
+    const uint64_t field_offset = offset_start + field->id() * sizeof(uint16_t);
+
+    if (field_offset >= vtable_offset + vtable_size) {
+      // This field_offset is too large for this vtable, so it must come from a
```

Comment:
I think we should still make a region for it "Unrecognized field from the future" or something like that.

I think the current implementation will call this padding?

Reply:
No, there is no empty space in this case. The binary was created with some earlier version of the schema and only has `N` fields. When processed with a newer version of the schema with  `N+1` fields we run into this situation.

Reply:
oh gotcha, I misread this. This is skipping fields that are in the schema and not in the binary. That makes sense. I confused that with the converse.

Still, what happens if there's a binary vtable with 10 fields but the schema only recognizes 4? I think the 6 fields of the vtable won't get `BinaryRegion`s since you make regions for all fields in the schema, rather than make regions for all fields in the binary.